README.md
odt ![Build Status](https://travis-ci.org/komola/odt.svg)
===

On Demand Thumbnail generator

Usage
-----

### Starting the handler

```
$ ./bin/odt --help

  Usage: odt [options]

  Options:

    -h, --help                                 output usage information
    -V, --version                              output the version number
    -i, --instances [instance number]          Number of handler instances. Default: Number of CPU cores
    -p, --port [port]                          Handler port. Default: 5000
    -w, --queue-port [interface port]          Queue interface port. Set to false to disable interface. Default: 4900
    -c, --config <config file>                 Specify a config file to load
    --original-storage-type <type>             Select original storage type.
    --original-storage-source-path <source>    Set original storage source path
    --original-storage-container <container>   Set original storage container name
    --original-storage-username <username>     Set original storage username
    --original-storage-password <password>     Set original storage password
    --original-storage-url <url>               Set original storage url
    --thumbnail-storage-type [type]            Select thumbnail storage type.
    --thumbnail-storage-source-path [source]   Set thumbnail storage source path
    --thumbnail-storage-container <container>  Set thumbnail storage container name
    --thumbnail-storage-username <username>    Set thumbnail storage username
    --thumbnail-storage-password <password>    Set thumbnail storage password
    --thumbnail-storage-url <url>              Set thumbnail storage url
    --statsd-host <host>                       Set statsd host
    --statsd-port <port>                       Set statsd port. Default 8125
    --statsd-prefix <prefix>                   Set statsd prefix.
    --redis-host <host>                        Set redis host. Default: localhost
    --redis-port <port>                        Set redis port. Default: 6379
    --redis-auth <auth>                        Set redis authentication password
    --secret <secret>                          Set a secret used for token generation
    --enable-token-only                        Enable token only access
```

### Starting the worker

```
$ ./bin/odt-worker --help

  Usage: odt-worker [options]

  Options:

    -h, --help                                 output usage information
    -V, --version                              output the version number
    -i, --instances [instance number]          Number of worker instances. Default: Number of CPU cores
    -c, --config <config file>                 Specify a config file to load
    --original-storage-type <type>             Select original storage type.
    --original-storage-source-path <source>    Set original storage source path
    --original-storage-container <container>   Set original storage container name
    --original-storage-username <username>     Set original storage username
    --original-storage-password <password>     Set original storage password
    --original-storage-url <url>               Set original storage url
    --thumbnail-storage-type [type]            Select thumbnail storage type. If not passed, will use the same as original storage.
    --thumbnail-storage-source-path [source]   Set thumbnail storage source path
    --thumbnail-storage-container <container>  Set thumbnail storage container name
    --thumbnail-storage-username <username>    Set thumbnail storage username
    --thumbnail-storage-password <password>    Set thumbnail storage password
    --thumbnail-storage-url <url>              Set thumbnail storage url
    --statsd-host <host>                       Set statsd host
    --statsd-port <port>                       Set statsd port. Default 8125
    --statsd-prefix <prefix>                   Set statsd prefix.
    --redis-host <host>                        Set redis host. Default: localhost
    --redis-port <port>                        Set redis port. Default: 6379
    --redis-auth <auth>                        Set redis authentication password
```

## Handler endpoints

When started, you can access images on the handler by requesting this url:

`http://localhost:5000/v1/400/300/image.jpg`

### Adding filters

You can define filters that should be added on top of the image. These filters
can be things like

* watermarks

The url for this type of requests is:

`http://localhost:5000/v1/400/300/filters:watermark(logo.png, 0, center)/image.jpg`

## Available filters

### Watermark
Valid parameters are:

* `watermark(*filename*, *x*, *y*, *opacity*)`
* `watermark(*filename*, *opacity*, tile)` tile the watermark over the image
* `watermark(*filename*, *opacity*, cover)` scale the watermark to cover the whole image
* `watermark(*filename*, *opacity*, center)` position the watermark in the center without resizing


debian/changelog
odt (1.0.0) unstable; urgency=low

  [ PrismaBox ]
  * Automatic build system complete

 -- PrismaBox <info@prismabox.de>  Mon, 11 Aug 2014 13:16:40 +0000


server.js
var program = require("commander");
var version = require("./package.json").version;
var _ = require("lodash");
var path = require("path");

program
  .version(version)
  .option('-i, --instances [instance number]', 'Number of handler instances. Default: Number of CPU cores', parseInt)
  .option('-p, --port [port]', 'Handler port. Default: 5000', parseInt)
  .option('-w, --queue-port [interface port]', 'Queue interface port. Set to false to disable interface. Default: 4900')
  .option('-c, --config <config file>', 'Specify a config file to load')

  .option('--original-storage-type <type>', 'Select original storage type.')
  .option('--original-storage-source-path <source>', 'Set original storage source path')

  .option('--original-storage-container <container>', 'Set original storage container name')
  .option('--original-storage-username <username>', 'Set original storage username')
  .option('--original-storage-password <password>', 'Set original storage password')
  .option('--original-storage-url <url>', 'Set original storage url')

  .option('--thumbnail-storage-type [type]', 'Select thumbnail storage type.')
  .option('--thumbnail-storage-source-path [source]', 'Set thumbnail storage source path')

  .option('--thumbnail-storage-container <container>', 'Set thumbnail storage container name')
  .option('--thumbnail-storage-username <username>', 'Set thumbnail storage username')
  .option('--thumbnail-storage-password <password>', 'Set thumbnail storage password')
  .option('--thumbnail-storage-url <url>', 'Set thumbnail storage url')

  .option('--statsd-host <host>', 'Set statsd host')
  .option('--statsd-port <port>', 'Set statsd port. Default 8125')
  .option('--statsd-prefix <prefix>', 'Set statsd prefix.')

  .parse(process.argv);

if (program.config) {
  program.config = path.resolve(program.config);
  config = require(program.config);

  _.extend(program, config);
}


var options = {
  instances: program.instances,
  handlerPort: program.port,
  queuePort: program.queuePort,
  statsdHost: program.statsdHost,
  statsdPort: program.statsdPort,
  statsdPrefix: program.statsdPrefix,
  storage: {}
};

if (program.originalStorageType === 'local') {
  options.storage.original = {
    type: 'local',
    sourcePath: program.originalStorageSourcePath
  };

  options.storage.thumbnail = {
    type: 'local',
    sourcePath: program.originalStorageSourcePath
  };
}

if (program.originalStorageType === 'swift') {
  options.storage.original = {
    type: 'swift',
    container: program.originalStorageContainer,
    username: program.originalStorageUsername,
    password: program.originalStoragePassword,
    url: program.originalStorageUrl
  };

  options.storage.thumbnail = {
    type: 'swift',
    container: program.originalStorageContainer,
    username: program.originalStorageUsername,
    password: program.originalStoragePassword,
    url: program.originalStorageUrl
  };
}



if (program.thumbnailStorageType === 'local') {
  options.storage.thumbnail = {
    type: 'local',
    sourcePath: program.thumbnailStorageSourcePath
  };
}

if (program.thumbnailStorageType === 'swift') {
  options.storage.thumbnail = {
    type: 'swift',
    container: program.thumbnailStorageContainer,
    username: program.thumbnailStorageUsername,
    password: program.thumbnailStoragePassword,
    url: program.thumbnailStorageUrl
  };
}

require("coffee-script/register");
require("./app/server")(options);


worker.js
var program = require("commander");
var version = require("./package.json").version;
var _ = require("lodash");
var path = require("path");

program
  .version(version)
  .option('-i, --instances [instance number]', 'Number of worker instances. Default: Number of CPU cores', parseInt)
  .option('-c, --config <config file>', 'Specify a config file to load')

  .option('--original-storage-type <type>', 'Select original storage type.')
  .option('--original-storage-source-path <source>', 'Set original storage source path')
  .option('--original-storage-container <container>', 'Set original storage container name')
  .option('--original-storage-username <username>', 'Set original storage username')
  .option('--original-storage-password <password>', 'Set original storage password')
  .option('--original-storage-url <url>', 'Set original storage url')

  .option('--thumbnail-storage-type [type]', 'Select thumbnail storage type. If not passed, will use the same as original storage.')
  .option('--thumbnail-storage-source-path [source]', 'Set thumbnail storage source path')

  .option('--thumbnail-storage-container <container>', 'Set thumbnail storage container name')
  .option('--thumbnail-storage-username <username>', 'Set thumbnail storage username')
  .option('--thumbnail-storage-password <password>', 'Set thumbnail storage password')
  .option('--thumbnail-storage-url <url>', 'Set thumbnail storage url')

  .option('--statsd-host <host>', 'Set statsd host')
  .option('--statsd-port <port>', 'Set statsd port. Default 8125')
  .option('--statsd-prefix <prefix>', 'Set statsd prefix.')

  .option('--redis-host <host>', 'Set redis host. Default: localhost')
  .option('--redis-port <port>', 'Set redis port. Default: 6379')
  .option('--redis-auth <auth>', 'Set redis authentication password')

  .parse(process.argv)
  ;

if (program.config) {
  program.config = path.resolve(program.config);
  config = require(program.config);

  _.extend(program, config);
}

var options = {
  instances: program.instances,
  handlerPort: program.port,
  queuePort: program.queuePort,
  statsdHost: program.statsdHost,
  statsdPort: program.statsdPort,
  statsdPrefix: program.statsdPrefix,
  redisHost: program.redisHost,
  redisPort: program.redisPort,
  redisAuth: program.redisAuth,
  storage: {}
};

if (program.originalStorageType === 'local') {
  options.storage.original = {
    type: 'local',
    sourcePath: program.originalStorageSourcePath
  };

  options.storage.thumbnail = {
    type: 'local',
    sourcePath: program.originalStorageSourcePath
  };
}

if (program.originalStorageType === 'swift') {
  options.storage.original = {
    type: 'swift',
    container: program.originalStorageContainer,
    username: program.originalStorageUsername,
    password: program.originalStoragePassword,
    url: program.originalStorageUrl
  };

  options.storage.thumbnail = {
    type: 'swift',
    container: program.originalStorageContainer,
    username: program.originalStorageUsername,
    password: program.originalStoragePassword,
    url: program.originalStorageUrl
  };
}



if (program.thumbnailStorageType === 'local') {
  options.storage.thumbnail = {
    type: 'local',
    sourcePath: program.thumbnailStorageSourcePath
  };
}

if (program.thumbnailStorageType === 'swift') {
  options.storage.thumbnail = {
    type: 'swift',
    container: program.thumbnailStorageContainer,
    username: program.thumbnailStorageUsername,
    password: program.thumbnailStoragePassword,
    url: program.thumbnailStorageUrl
  };
}

options.isWorker = true;

require("coffee-script/register");
require("./app/server")(options);


odt.js
var program = require("commander");
var version = require("./package.json").version;
var _ = require("lodash");
var path = require("path");

program
  .version(version)
  .option('-i, --instances [instance number]', 'Number of handler instances. Default: Number of CPU cores', parseInt)
  .option('-p, --port [port]', 'Handler port. Default: 5000', parseInt)
  .option('-w, --queue-port [interface port]', 'Queue interface port. Set to false to disable interface. Default: 4900')
  .option('-c, --config <config file>', 'Specify a config file to load')

  .option('--original-storage-type <type>', 'Select original storage type.')
  .option('--original-storage-source-path <source>', 'Set original storage source path')

  .option('--original-storage-container <container>', 'Set original storage container name')
  .option('--original-storage-username <username>', 'Set original storage username')
  .option('--original-storage-password <password>', 'Set original storage password')
  .option('--original-storage-url <url>', 'Set original storage url')

  .option('--thumbnail-storage-type [type]', 'Select thumbnail storage type.')
  .option('--thumbnail-storage-source-path [source]', 'Set thumbnail storage source path')

  .option('--thumbnail-storage-container <container>', 'Set thumbnail storage container name')
  .option('--thumbnail-storage-username <username>', 'Set thumbnail storage username')
  .option('--thumbnail-storage-password <password>', 'Set thumbnail storage password')
  .option('--thumbnail-storage-url <url>', 'Set thumbnail storage url')

  .option('--statsd-host <host>', 'Set statsd host')
  .option('--statsd-port <port>', 'Set statsd port. Default 8125')
  .option('--statsd-prefix <prefix>', 'Set statsd prefix.')

  .option('--redis-host <host>', 'Set redis host. Default: localhost')
  .option('--redis-port <port>', 'Set redis port. Default: 6379')
  .option('--redis-auth <auth>', 'Set redis authentication password')

  .option('--secret <secret>', 'Set a secret used for token generation')
  .option('--enable-token-only', 'Enable token only access')

  .parse(process.argv);

if (program.config) {
  program.config = path.resolve(program.config);
  config = require(program.config);

  _.extend(program, config);
}


var options = {
  instances: program.instances,
  handlerPort: program.port,
  queuePort: program.queuePort,
  statsdHost: program.statsdHost,
  statsdPort: program.statsdPort,
  statsdPrefix: program.statsdPrefix,
  redisHost: program.redisHost,
  redisPort: program.redisPort,
  redisAuth: program.redisAuth,
  storage: {},
  enableTokenOnly: program.enableTokenOnly || false,
  secret: program.secret || ""
};

if (program.originalStorageType === 'local') {
  options.storage.original = {
    type: 'local',
    sourcePath: program.originalStorageSourcePath
  };

  options.storage.thumbnail = {
    type: 'local',
    sourcePath: program.originalStorageSourcePath
  };
}

if (program.originalStorageType === 'swift') {
  options.storage.original = {
    type: 'swift',
    container: program.originalStorageContainer,
    username: program.originalStorageUsername,
    password: program.originalStoragePassword,
    url: program.originalStorageUrl
  };

  options.storage.thumbnail = {
    type: 'swift',
    container: program.originalStorageContainer,
    username: program.originalStorageUsername,
    password: program.originalStoragePassword,
    url: program.originalStorageUrl
  };
}

if (program.thumbnailStorageType === 'local') {
  options.storage.thumbnail = {
    type: 'local',
    sourcePath: program.thumbnailStorageSourcePath
  };
}

if (program.thumbnailStorageType === 'swift') {
  options.storage.thumbnail = {
    type: 'swift',
    container: program.thumbnailStorageContainer,
    username: program.thumbnailStorageUsername,
    password: program.thumbnailStoragePassword,
    url: program.thumbnailStorageUrl
  };
}

require("coffee-script/register");
require("./app/server")(options);
